src/usr.bin/locate/locate/locate.1
.\"	$OpenBSD: locate.1,v 1.22 2004/04/08 20:08:59 jmc Exp $
.\"
.\" Copyright (c) 1995 Wolfram Schneider <wosch@FreeBSD.org>. Berlin.
.\" Copyright (c) 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)locate.1	8.1 (Berkeley) 6/6/93
.\"	$Id$
.\"
.Dd June 6, 1993
.Dt LOCATE 1
.Os
.Sh NAME
.Nm locate
.Nd find filenames quickly
.Sh SYNOPSIS
.Nm locate
.Op Fl cimSs
.Op Fl d Ar database
.Op Fl l Ar limit
.Ar pattern Op Ar ...
.Sh DESCRIPTION
The
.Nm
utility searches a database for all pathnames which match the specified
.Ar pattern .
The database is recomputed periodically (usually weekly or daily),
and contains the pathnames
of all files which are publicly accessible.
.Pp
Shell globbing and quoting characters
.Pf ( Ql * ,
.Ql \&? ,
.Ql \e ,
.Ql [ ,
and
.Ql \&] )
may be used in
.Ar pattern ,
although they will have to be escaped from the shell.
Preceding any character with a backslash
.Pq Ql \e
eliminates any special meaning which it may have.
The matching differs in that no characters must be matched explicitly,
including slashes
.Pq Ql / .
.Pp
As a special case, a pattern containing no globbing characters
.Pq Dq foo
is matched as though it were
.Dq *foo* .
.Pp
Historically,
.Nm
stores only characters between 32 and 127.
The current implementation stores all characters except newline
.Pq Ql \en
and
NUL
.Pq Ql \e0 .
The 8-bit character support does not waste extra space for
plain
.Tn ASCII
file names.
Characters less than 32 or greater than 127 are stored as 2 bytes.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c
Suppress normal output; instead print a count of matching file names.
.It Fl d Ar database
Search in
.Ar database
instead of the default file name database.
Multiple
.Fl d
options are allowed.
Each additional
.Fl d
option adds the specified database to the list
of databases to be searched.
.Pp
.Ar database
may be a colon-separated list of databases.
A single colon is a reference to the default database.
.Pp
.Dl $ locate -d $HOME/lib/mydb: foo
.Pp
will first search for the string
.Dq foo
in
.Pa $HOME/lib/mydb
and then in
.Pa /var/db/locate.database .
.Pp
.Dl $ locate -d $HOME/lib/mydb::/cdrom/locate.database foo
.Pp
will first search for the string
.Dq foo
in
.Pa $HOME/lib/mydb
and then in
.Pa /var/db/locate.database
and then in
.Pa /cdrom/locate.database .
.Pp
.Dl $ locate -d db1 -d db2 -d db3 pattern
.Pp
is the same as
.Pp
.Dl $ locate -d db1:db2:db3 pattern
.Pp
or
.Pp
.Dl $ locate -d db1:db2 -d db3 pattern
.Pp
If
.Ql \-
is given as the
.Ar database
name, standard input will be read instead.
For example, you can compress your database
and use:
.Pp
.Dl $ zcat database.gz | locate -d - pattern
.Pp
This might be useful on machines with a fast CPU, little RAM and slow I/O.
.Sy Note:
You can only use
.Em one
pattern for stdin.
.It Fl i
Ignore case distinctions in both the pattern and the database.
.It Fl l Ar number
Limit output to
.Ar number
of file names and exit.
.It Fl m
Use
.Xr mmap 2
instead of the
.Xr stdio 3
library.
This is the default behavior.
Usually faster in most cases.
.It Fl S
Print some statistics about the database and exit.
.It Fl s
Use the
.Xr stdio 3
library instead of
.Xr mmap 2 .
.El
.Sh ENVIRONMENT
.Bl -tag -width LOCATE_PATH -compact
.It Ev LOCATE_PATH
Path to the locate database if set and not empty; ignored if the
.Fl d
option was specified.
.El
.Sh FILES
.Bl -tag -width /usr/libexec/locate.updatedb -compact
.It Pa /etc/weekly
script that starts the database rebuild
.It Pa /usr/libexec/locate.updatedb
script to update the locate database
.It Pa /var/db/locate.database
locate database
.El
.Sh SEE ALSO
.Xr find 1 ,
.Xr fnmatch 3 ,
.Xr locate.updatedb 8 ,
.Xr weekly 8
.Rs
.%A Woods, James A.
.%D 1983
.%T "Finding Files Fast"
.%J ";login"
.%V 8:1
.%P pp. 8-10
.Re
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.4 .
.Sh BUGS
.Nm
may fail to list some files that are present, or may
list files that have been removed from the system.
This is because
.Nm
only reports files that are present in a periodically reconstructed
database (typically rebuilt once a week by the
.Xr weekly 8
script).
Use
.Xr find 1
to locate files that are of a more transitory nature.
.Pp
The
.Nm
database is built by user
.Dq nobody
using
.Xr find 1 .
This will
skip directories which are not readable by user
.Dq nobody ,
group
.Dq nobody ,
or
the world.
E.g., if your home directory is not world-readable, your files will
.Em not
appear in the database.
.Pp
The
.Nm
database is not byte order independent.
It is not possible
to share the databases between machines with different byte order.
The current
.Nm
implementation understands databases in host byte order or
network byte order.
So a little-endian machine can't understand
a locate database which was built on a big-endian machine.


src/usr.bin/locate/locate/locate.c
/*
 *	$OpenBSD: locate.c,v 1.18 2004/04/08 20:08:59 jmc Exp $
 *
 * Copyright (c) 1995 Wolfram Schneider <wosch@FreeBSD.org>. Berlin.
 * Copyright (c) 1989, 1993
 *      The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * James A. Woods.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *      $Id$
 */

#include <sys/cdefs.h>
__COPYRIGHT("@(#) Copyright (c) 1995-1996 Wolfram Schneider, Berlin.\n\
@(#) Copyright (c) 1989, 1993\n\
        The Regents of the University of California.  All rights reserved.\n");
__SCCSID("@(#)locate.c    8.1 (Berkeley) 6/6/93");
__RCSID("$MirOS$");

/*
 * Ref: Usenix ;login:, Vol 8, No 1, February/March, 1983, p. 8.
 *
 * Locate scans a file list for the full pathname of a file given only part
 * of the name.  The list has been processed with with "front-compression"
 * and bigram coding.  Front compression reduces space by a factor of 4-5,
 * bigram coding by a further 20-25%.
 *
 * The codes are:
 *
 *      0-28    likeliest differential counts + offset to make nonnegative
 *      30      switch code for out-of-range count to follow in next word
 *      31      an 8 bit char followed
 *      128-255 bigram codes (128 most common, as determined by 'updatedb')
 *      32-127  single character (printable) ascii residue (ie, literal)
 *
 * A novel two-tiered string search technique is employed:
 *
 * First, a metacharacter-free subpattern and partial pathname is matched
 * BACKWARDS to avoid full expansion of the pathname list.  The time savings
 * is 40-50% over forward matching, which cannot efficiently handle
 * overlapped search patterns and compressed path residue.
 *
 * Then, the actual shell glob-style regular expression (if in this form) is
 * matched against the candidate pathnames using the slower routines provided
 * in the standard 'find'.
 */

#include <sys/param.h>
#include <ctype.h>
#include <err.h>
#include <fnmatch.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef MMAP
#  include <sys/types.h>
#  include <sys/stat.h>
#  include <sys/mman.h>
#  include <fcntl.h>
#endif


#ifdef sun
#include <netinet/in.h> /* SunOS byteorder(3) htohl(3) */
#endif

#include "locate.h"
#include "pathnames.h"

#ifdef DEBUG
#  include <sys/time.h>
#  include <sys/types.h>
#  include <sys/resource.h>
#endif

char *path_fcodes;      /* locate database */
int f_mmap;             /* use mmap */
int f_icase;            /* ignore case */
int f_stdin;            /* read database from stdin */
int f_statistic;        /* print statistic */
int f_silent;           /* suppress output, show only count of matches */
int f_limit;            /* limit number of output lines, 0 == infinite */
u_int counter;          /* counter for matches [-c] */


void    usage(void);
void    statistic(FILE *, char *);
void    fastfind(FILE *, char *, char *);
void    fastfind_icase(FILE *, char *, char *);
void    fastfind_mmap(char *, caddr_t, int, char *);
void    fastfind_mmap_icase(char *, caddr_t, int, char *);
void	search_mmap(char *, char **);
void	search_fopen(char *, char **);
unsigned long cputime(void);

extern char     **colon(char **, char*, char*);
extern void     print_matches(u_int);
extern int      getwm(caddr_t);
extern int      getwf(FILE *);
extern u_char   *tolower_word(u_char *);
extern int	check_bigram_char(int);
extern char 	*patprep(char *);

extern char     *optarg;
extern int      optind;


int
main(int argc, char *argv[])
{
	int ch;
	char **dbv = NULL;
#ifdef MMAP
	f_mmap = 1;		/* mmap is default */
#endif
#ifndef __MirBSD__
	(void) setlocale(LC_ALL, "");
#endif

	while ((ch = getopt(argc, argv, "Scd:il:ms")) != -1)
		switch (ch) {
		case 'S':	/* statistic lines */
			f_statistic = 1;
			break;
		case 'l': /* limit number of output lines, 0 == infinite */
			f_limit = atoi(optarg);
			break;
		case 'd':	/* database */
			dbv = colon(dbv, optarg, _PATH_FCODES);
			break;
		case 'i':	/* ignore case */
			f_icase = 1;
			break;
		case 'm':	/* mmap */
#ifdef MMAP
			f_mmap = 1;
#else
			(void)fprintf(stderr, "mmap(2) not implemented\n");
#endif
			break;
		case 's':	/* stdio lib */
			f_mmap = 0;
			break;
		case 'c': /* suppress output, show only count of matches */
			f_silent = 1;
			break;
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	/* to few arguments */
	if (argc < 1 && !(f_statistic))
		usage();

	/* no (valid) database as argument */
	if (dbv == NULL || *dbv == NULL) {
		/* try to read database from environment */
		if ((path_fcodes = getenv("LOCATE_PATH")) == NULL ||
		    *path_fcodes == '\0')
			/* use default database */
			dbv = colon(dbv, _PATH_FCODES, _PATH_FCODES);
		else		/* $LOCATE_PATH */
			dbv = colon(dbv, path_fcodes, _PATH_FCODES);
	}

	if (f_icase && UCHAR_MAX < 4096) /* init tolower lookup table */
		for (ch = 0; ch < UCHAR_MAX + 1; ch++)
			myctype[ch] = tolower(ch);

	/* foreach database ... */
	while ((path_fcodes = *dbv) != NULL) {
		dbv++;

		if (!strcmp(path_fcodes, "-"))
			f_stdin = 1;
		else
			f_stdin = 0;

#ifndef MMAP
		f_mmap = 0;	/* be paranoid */
#endif
		if (!f_mmap || f_stdin || f_statistic)
			search_fopen(path_fcodes, argv);
		else
			search_mmap(path_fcodes, argv);
	}

	if (f_silent)
		print_matches(counter);
	exit(0);
}


void
search_fopen(char *db, char **s)
{
	FILE *fp;
#ifdef DEBUG
	long t0;
#endif

	/* can only read stdin once */
	if (f_stdin) {
		fp = stdin;
		if (*(s+1) != NULL) {
			(void)fprintf(stderr,
			    "read database from stdin, use only");
			(void)fprintf(stderr, " `%s' as pattern\n", *s);
			*(s+1) = NULL;
		}
	}
	else if ((fp = fopen(path_fcodes, "r")) == NULL)
		err(1,  "`%s'", path_fcodes);

	/* count only chars or lines */
	if (f_statistic) {
		statistic(fp, path_fcodes);
		(void)fclose(fp);
		return;
	}

	/* foreach search string ... */
	while (*s != NULL) {
#ifdef DEBUG
		t0 = cputime();
#endif
		if (!f_stdin &&
		    fseek(fp, (long)0, SEEK_SET) == -1)
			err(1, "fseek to begin of ``%s''", path_fcodes);

		if (f_icase)
			fastfind_icase(fp, *s, path_fcodes);
		else
			fastfind(fp, *s, path_fcodes);
#ifdef DEBUG
		(void)fprintf(stderr, "fastfind %ld ms\n", cputime () - t0);
#endif
		s++;
	}
	(void)fclose(fp);
}

#ifdef MMAP
void
search_mmap(char *db, char **s)
{
	struct stat sb;
	int fd;
	caddr_t p;
	off_t len;
#ifdef DEBUG
	long t0;
#endif
	if ((fd = open(path_fcodes, O_RDONLY)) == -1 ||
	    fstat(fd, &sb) == -1)
		err(1, "`%s'", path_fcodes);
	len = sb.st_size;

	if ((p = mmap((caddr_t)0, (size_t)len, PROT_READ, MAP_SHARED,
	    fd, (off_t)0)) == MAP_FAILED)
		err(1, "mmap ``%s''", path_fcodes);

	/* foreach search string ... */
	while (*s != NULL) {
#ifdef DEBUG
		t0 = cputime();
#endif
		if (f_icase)
			fastfind_mmap_icase(*s, p, (int)len, path_fcodes);
		else
			fastfind_mmap(*s, p, (int)len, path_fcodes);
#ifdef DEBUG
		(void)fprintf(stderr, "fastfind %ld ms\n", cputime () - t0);
#endif
		s++;
	}

	if (munmap(p, (size_t)len) == -1)
		warn("munmap %s", path_fcodes);

	(void)close(fd);
}
#endif /* MMAP */

#ifdef DEBUG
unsigned long
cputime(void)
{
	struct rusage rus;

	getrusage(0, &rus);
	return(rus.ru_utime.tv_sec * 1000 + rus.ru_utime.tv_usec / 1000);
}
#endif /* DEBUG */

void
usage(void)
{
	(void)fprintf(stderr, "usage: locate [-cimSs] [-d database] ");
	(void)fprintf(stderr, "[-l limit] pattern ...\n");
	(void)fprintf(stderr, "default database: `%s' or $LOCATE_PATH\n",
	    _PATH_FCODES);
	exit(1);
}


/* load fastfind functions */

/* statistic */
/* fastfind_mmap, fastfind_mmap_icase */
#ifdef MMAP
#undef FF_MMAP
#undef FF_ICASE

#define FF_MMAP
#include "fastfind.c"
#define FF_ICASE
#include "fastfind.c"
#endif /* MMAP */

/* fopen */
/* fastfind, fastfind_icase */
#undef FF_MMAP
#undef FF_ICASE
#include "fastfind.c"
#define FF_ICASE
#include "fastfind.c"
